### ISSUE TITLE
ERD: Improve auto layout

### ISSUE THREAD

**Issue by adityatoshniwal** [pgadmin-org/pgadmin4#6592]
**Describe the solution you'd like**
Currently, the auto layout in ERD places the nodes correctly but links are:
1. Not always right angled.
2. Goes behind the nodes.
3. Does not switch sides to get best link route.

This can be improved by using a better algorithm.

**Comment by wayland** [comment]
![Screenshot_20230825_164341](https://github.com/pgadmin-org/pgadmin4/assets/60692/57d94ca0-1bda-4d22-b048-c57f8698885e)
An example of the lines not going anywhere very sensible.  I'd personally call this a bugfix rather than a feature, though it's an aesthetic bug rather than a functional one.  

HTH

**Comment by wayland** [comment]
May also be related to https://github.com/pgadmin-org/pgadmin4/issues/5199

**Comment by morrissimo** [comment]
seeing identical behavior on pgadmin 7.8 on Mac - after generating an ERD for an existing database, not only do the relationship lines look pretty wild, but they also have anchor points that don't reset and aren't movable/editable.

Note that this seems to be the only issue preventing an autogenerated ERD from being useful - everything else looks pretty good!

**Comment by raprocks** [comment]
Any progress on this or can someone point me in the direction where i can start working on it?

**Comment by adityatoshniwal** [comment]
Hi @raprocks,
Right now, we're failing to find a relevant layout algorithm. Basically, a layout algo will help in getting the positions of nodes and links. We're using dagre currently https://www.npmjs.com/package/dagre

**Comment by carlo-salinari** [comment]
For comparison, I'm attaching an auto-generated er diagram of the same (very simple) db done with pgAdmin4 v7.8 and a proprietary software (Navicat Data Modeler 3).

![image](https://github.com/pgadmin-org/pgadmin4/assets/816409/fcd62908-4a4a-4168-ac6f-4f58da8bf4f4)

![Diagram](https://github.com/pgadmin-org/pgadmin4/assets/816409/cfc2375e-6b7c-4e5b-a076-c39bee388223)

I'm no expert, but maybe a first step in the right direction could be switching to [RightAngleLink](https://github.com/projectstorm/react-diagrams/blob/master/diagrams-demo-gallery/demos/demo-right-angles-routing/index.tsx)s as per `@projectstorm/react-diagrams` [demo](https://projectstorm.cloud/react-diagrams/?path=/story/advanced-usage--right-angles-routing).

(I'm with @wayland in that this looks more like a bugfix than a feature-request. I do recognize that automatic laying-out is very complex to get right, and will probably require many iterations).

**Comment by adityatoshniwal** [comment]
@carlo-salinari  The major difference between pgAdmin ERD and the ERD which you shared is - The links in pgAdmin start from the actual column used, and not just randomly sticking the link anywhere.
If we're OK to sacrifice this then a solution can be worked on.

**Comment by carlo-salinari** [comment]
@adityatoshniwal I believe from the end-user perspective the major difference is in terms of _clarity_, so to speak.

And that seems to come from the routing around (and not behind) the tables, better use of spacing and grouping, predictability of the links, overall less graphical noise from the background and better color contrast.

The direct connection to the column certainly complicates routing a bit, but it could probably be maintained. It would be interesting to find comparisons.

**Comment by adityatoshniwal** [comment]
@carlo-salinari I had tried to find a suitable algorithm to keep links on column and auto layout. But unfortunately didn't find. I think at some point we'll have to let it go in favour of better clarity. In future if we find a suitable algo then we can link back to columns as an improvement.

**Comment by carlo-salinari** [comment]
@adityatoshniwal You mean let go of the link-to-key constraint? That's something to experiment with, but I don't think that's the crux of the problem here.

Automatic graph layout is a notoriously hard problem: there is no clearly understood metric to evaluate the results and the complexity varies wildly with graph input size and constraints number.

These two aspects will keep coming up while looking for a solution (even a partial one).

It's unlikely that we'll be able to find a ready-made robust algorithm able to handle arbitrarily sized schemas right away, even removing the link-to-key constraint.

A more pragmatic approach would be to work step-wise, i.e.:

1. establish thresholds (small, medium, large schemas) and gather sample schemas
2. work towards making automatic layout of small/medium schemas mostly work right out of the box (experiment with and fine-tune some classic force directed/network flow algorithm)
3. make fixing schemas by hand as painless as possible (right now it's a pretty frustrating experience)

This seems more or less the approach currently followed by all other similar applications.

And eventually some magic machine-learning solution will materialize, capturing all the aesthetic requirements :)

**Comment by carlo-salinari** [comment]
This is a sample of an auto-generated erd graph with maintained link-to-key constraints (https://www.softwareideas.net/en/erd-auto-generator).

![erd-auto-generator](https://github.com/pgadmin-org/pgadmin4/assets/816409/88b16b23-98f8-4428-a6b3-e8428baa099a)

- layout is orthogonal
- links route around tables
- bends variance appear uniform
- crossings are marked graphically
- not much in the way of compaction

**Comment by carlo-salinari** [comment]
Only javascript library I could find tackling erd diagrams decently: https://gojs.net/latest/samples/entityRelationship.html

![image](https://github.com/pgadmin-org/pgadmin4/assets/816409/60d225e1-c07e-460d-bf68-abe6752b0afe)

Unfortunately, I don't think their license is compatible with pgAdmin's.

**Comment by wayland** [comment]
Other people (eg. DataWharf) seem to be using maxGraph (basically mxGraph, but gets updates).  They're under an Apache-style license; no idea about compatibility.  

https://github.com/maxGraph/maxGraph

HTH,

**Comment by javinosearas** [comment]
It would be very helpful to add a feature to manually move the nodes of the lines manually by clicking and dragging. That way it would be a lot easier to tidy up an auto generated ERD.  The ability to manually insert and delete nodes on a line segment would also come in handy.

**Comment by RieBi** [comment]
It seems like orthogonal lines is the way

**Comment by adityatoshniwal** [comment]
@RieBi ERD tool lines works fine when building from scratch. The issue is with auto layout. I will go through all the suggestions and see how we can fix it.

**Comment by vishnupeas** [comment]
This is clearly a 'bug' not a 'feature' from end user perspective. Also this issue should get a better priority imo.

**Comment by adityatoshniwal** [comment]
> This is clearly a 'bug' not a 'feature' from end user perspective. Also this issue should get a better priority imo.

Feel free to raise a PR if you have a solution. I've spend a lot of time in past, may be there is one available today.

**Comment by adityatoshniwal** [comment]
Point no.1 and 3 will be fixed with my PR here - https://github.com/pgadmin-org/pgadmin4/pull/8167.
For point no. 2, I spent a lot of time to find a apt. path finding algorithm.
I tried A-star but it becomes very very slow for large diagrams - and produces links with lot of turns. If anybody happens to know a good path finding lib then we can use that. Intention is not to get the best optimal path but to get a path with least turns without going below nodes.
